draft-ietf-ppm-dap.md

```diff
@@ -1200,34 +1182,120 @@ enum {
+splitting each report into report shares, one for each Aggregator. The Leader
+and Helper then run the aggregate initialization flow to accomplish two tasks:
+
+1. Recover and determine which input report shares are valid.
+1. For each valid report share, initialize the VDAF preparation process (see
+   {{Section 5.2 of !VDAF}}).
+
 The Leader and Helper initialization behavior is detailed below.
 
 #### Leader Initialization {#leader-init}
 
 The Leader begins the aggregate initialization phase with the set of candidate
-report shares as follows:
+reports as follows:
 
 1. Generate a fresh AggregationJobID. The ID value MUST be unique within the
    scope of the corresponding DAP task.
 1. Decrypt the input share for each report share as described in
    {{input-share-decryption}}.
 1. Check that the resulting input share is valid as described in
    {{input-share-validation}}.
-1. Initialize VDAF preparation as described in {{input-share-prep}}.
 
-If any step invalidates the report share, the Leader removes the report share
-from the set of candidate reports. Once the leader has initialized this state
-for all valid candidate report shares, it creates an `AggregationJobInitReq`
-message for each Helper to initialize the preparation of this candidate set. The
-`AggregationJobInitReq` message is structured as follows:
+If any step invalidates the report, the Leader rejects the report and removes
+it from the set of candidate reports.
+
+Next, for each report the Leader executes the following procedure:
+
+~~~
+state = VDAF.ping_pong_start(vdaf_verify_key,
+                             True,
+                             agg_param,
+                             report_id,
+                             public_share,
+                             plaintext_input_share.payload)
+if state != Rejected():
+  (state, outbound) = VDAF.ping_pong_req(agg_param, state, None)
+~~~
+
+where:
+
+* `vdaf_verify_key` is the VDAF verification key for the task
+* `agg_param` is the VDAF aggregation parameter provided by the Collector (see
+  {{collect-flow}})
+* `report_id` is the report ID
+* `public_share` is the report's public share
+* `plaintext_input_share` is the Leader's `PlaintextInputShare`
+
+The methods are defined in {{!VDAF}}. This process determines the initial
+per-report `state`, as well as the initial `outbound` message to send to the
+Helper. (These are coalesced into a single HTTP request to the Helper as
+described below.) If `state == Rejected()`, then the report is rejected and
+removed from the set of candidate reports.
+
+[[OPEN ISSUE: update references to the ping-pong API, once a VDAF draft is
+published which describes it.]]
+
+Next, for each candidate report the Leader constructs a `PrepareStep` message
+of type "initialize":
 
 ~~~
 struct {
-  ReportMetadata report_metadata;
-  opaque public_share<0..2^32-1>;
-  HpkeCiphertext encrypted_input_share;
-} ReportShare;
+  PrepareStepState prepare_step_state = 0; /* initialize */
```

Comment:
Both are valid I think. I'm fine with this change, but I'd err on the side of being consistent with the VDAF spec. There I'm not sure we want to change this since the pseudocode is Python and this would look slightly less Pythonic.

Reply:
You're going to quite correctly accuse me of being a pedant, but this particular snippet is not a Python description of calling a VDAF method but rather TLS presentation language explaining how to construct a wire message. If we look at [an example from RFC 8446](https://datatracker.ietf.org/doc/html/rfc8446#appendix-B.1):

```
enum {
          invalid(0),
          change_cipher_spec(20),
          alert(21),
          handshake(22),
          application_data(23),
          heartbeat(24),  /* [RFC 6520](https://datatracker.ietf.org/doc/html/rfc6520) */
          (255)
      } ContentType;

      <snip>

      struct {
          ContentType opaque_type = application_data; /* 23 */
          ProtocolVersion legacy_record_version = 0x0303; /* TLS v1.2 */
          uint16 length;
          opaque encrypted_record[TLSCiphertext.length];
      } TLSCiphertext;
```

Reply:
We'll have to revisit this in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/472 anyway, so @branlwyd I think you should make the call on whether you want to update these in this PR. It's perfectly readable either way.

Reply:
I like specifying the name, I don't think it's any less clear (and implementors will probably be writing a name rather than a number) & it's supported by existing notation.

Reply:
@tgeoghegan I'm not sure what you mean? In VDAF we currently write:
```python
﻿﻿﻿    elif inbound.type == 1: # continue
```
If we instead wrote something like
```python
    elif inbound.type == continue:
```
then we'd need to define `continue` as a constant somewhere. Which, you know, we could do.

Reply:
The snippet that we're discussing in this thread is:

```
~~~
struct {
  PrepareStepState prepare_step_state = 0; /* initialize */
  ReportShare report_share = helper_report_share;
  opaque payload<0..2^32-1> = outbound;
} PrepareStep;
~~~
```

That's not Python, it's RFC 8446 presentation language, so Python syntax is irrelevant here.